src/server.rs

```
@@ -12,48 +29,207 @@ pub struct ServerConfig {
+            status_err!("failed to build descriptor service: {}", err);
+            std::process::exit(1)
+        });
+    let config = APP.config();
+    let server_config = load_server_config(&config).await.unwrap_or_else(|err| {
+        status_err!("failed to load server config: {}", err);
+        std::process::exit(1)
+    });
+
+    info!("listening on {}", server_config.address);
+    if let Err(err) = tonic::transport::Server::builder()
+        .tls_config(server_config.tls_config.expect("must have tls config"))
+        .unwrap_or_else(|err| {
+            status_err!("failed to configure server for tls: {}", err);
+            std::process::exit(1)
+        })
+        .add_service(ContractCallServiceServer::new(CorkHandler))
+        .add_service(proto_descriptor_service)
+        .serve_with_shutdown(server_config.address, cancellation_token.cancelled())
+        .await
+    {
+        status_err!("server error: {}", err);
+        std::process::exit(1)
+    }
+}
+
+pub(crate) async fn auth_config(
+    config: &Arc<StewardConfig>,
+) -> Result<rustls::ServerConfig, Error> {
+    let mut trust_store = rustls::RootCertStore::empty();
+    let cache = PUBLISHER_TRUST_STATE_CACHE.read().await;
+    for td in cache.values() {
+        trust_store
+            .add_pem_file(&mut td.publisher.ca_cert.as_bytes())
+            .unwrap();
+    }
+
+    let client_auth = rustls::AllowAnyAuthenticatedClient::new(trust_store);
+
+    let cert = tokio::fs::read(&config.server.server_cert_path).await?;
+    let key = tokio::fs::read(&config.server.server_key_path).await?;
+    let cert: Vec<rustls::Certificate> = rustls_pemfile::certs(&mut cert.as_slice())
+        .unwrap()
+        .iter()
+        .map(|v| rustls::Certificate(v.clone()))
+        .collect();
+    let key = match rustls_pemfile::read_one(&mut key.as_slice())
+        .expect("cannot parse private key .pem file")
+    {
+        Some(rustls_pemfile::Item::RSAKey(key)) => rustls::PrivateKey(key),
```

I don't think we need to support RSA keys, we expect everyone to use EC keys. Is this a case of having to match every possible outcome? We might want to return an error.

---

Wasn't thinking. I'm a little confused by the PKCS8 variant; isn't that just a syntax for storing keys in the file? Do we need to determine if the underlying key is RSA or EC if the PKCS8 variant is hit for some reason?